build-llvm.py

```diff
@@ -0,0 +1,438 @@
+                        "--no-pull",
+                        help=textwrap.dedent("""\
+                        By default, the script always updates the LLVM repo before building. This prevents
+                        that, which can be helpful during something like bisecting.
+
+                        """),
+                        action="store_true")
+    parser.add_argument("-p",
+                        "--projects",
+                        help=textwrap.dedent("""\
+                        Currently, the script only enables the clang, compiler-rt, and lld folders in LLVM. If
+                        you would like to override this, you can use this parameter and supply a list that is
+                        supported by LLVM_ENABLE_PROJECTS.
+
+                        See step #5 here: https://llvm.org/docs/GettingStarted.html#getting-started-quickly-a-summary
+
+                        Example: -p \"clang;lld;libcxx\"
+
+                        """),
+                        type=str,
+                        default="clang;lld;compiler-rt")
+    parser.add_argument("-t",
+                        "--targets",
+                        help=textwrap.dedent("""\
+                        LLVM is multitargeted by default. Currently, this script only enables the arm32, aarch64,
+                        powerpc, and x86 backends because that's what the Linux kernel is currently concerned with.
+                        If you would like to override this, you can use this parameter and supply a list that is
+                        supported by LLVM_TARGETS_TO_BUILD: https://llvm.org/docs/CMake.html#llvm-specific-variables
+
+                        Example: -t "AArch64;X86"
+
+                        """),
+                        type=str,
+                        default="AArch64;ARM;PowerPC;X86")
+    return parser.parse_args()
+
+
+def linker_test(cc, ld):
+    """
+    Test to see if the supplied ld value will work with cc -fuse=ld
+    :param cc: A working C compiler to compile the test program
+    :param ld: A linker to test -fuse=ld against
+    :return: 0 if the linker supports -fuse=ld, 1 otherwise
+    """
+    echo = subprocess.Popen(['echo', 'int main() { return 0; }'],
+                            stdout=subprocess.PIPE)
+    cc_call = subprocess.Popen(
+        [cc, '-fuse-ld=' + ld, '-o', '/dev/null', '-x', 'c', '-'],
+        stdin=echo.stdout,
+        stderr=subprocess.DEVNULL)
+    cc_call.communicate()
+    return cc_call.returncode
+
+
+def check_cc_ld_variables():
+    """
+    Sets the cc, cxx, and ld variables, which will be passed to cmake
+    :return: A tuple of valid cc, cxx, ld values that can be used to compile LLVM
+    """
+    utils.print_header("Checking CC and LD")
+    cc, linker, ld = None, None, None
+    # If the user didn't supply a C compiler, try to find one
+    if 'CC' not in os.environ:
+        possible_compilers = ['clang-9', 'clang-8', 'clang-7', 'clang', 'gcc']
```

Comment:
Bikeshedding: if the branches are swapped, it may be easier to read the indentation and the negated comment (and no need for `not` either).